src/main/java/io/conduit/DefaultDestinationStream.java

```diff
@@ -47,32 +49,77 @@ public void onNext(Destination.Run.Request request) {
         } catch (Exception e) {
             logger.error("Couldn't write record.", e);
             responseObserver.onError(
-                Status.INTERNAL
-                    .withDescription("couldn't write record: " + e.getMessage())
-                    .withCause(e)
-                    .asException()
+                    Status.INTERNAL
+                            .withDescription("couldn't write record: " + e.getMessage())
+                            .withCause(e)
+                            .asException()
             );
         }
     }
 
     private Destination.Run.Response responseWith(ByteString position) {
         return Destination.Run.Response
-            .newBuilder()
-            .setAckPosition(position)
-            .build();
+                .newBuilder()
+                .setAckPosition(position)
+                .build();
     }
 
     private void doWrite(Record rec) {
-        logger.info("DO WRITE..");
-        logger.info(rec.toString());
-        // todo write into S3
+        switch (rec.getOperation()) {
+            case OPERATION_CREATE, OPERATION_SNAPSHOT:
+                insertRecord(rec);
+                break;
+            case OPERATION_UPDATE:
+                logger.info("Updates are not supported yet.");
+                break;
+            case OPERATION_DELETE:
+                deleteRecord(rec);
+                break;
+            default:
+                break;
+        }
+    }
+
+    @SneakyThrows
+    private void insertRecord(Record rec) {
+        Objects.requireNonNull(rec, "record is null");
+        var schema = spark.read().table(tableName).schema();
+
+        // todo handle structured data as well
+        String afterString = rec.getPayload().getAfter().getRawData().toStringUtf8();
+
+        Dataset<Row> data = spark.read()
+                .schema(schema)
+                .json(spark.createDataset(List.of(afterString), Encoders.STRING()));
+
+        data.write()
+                .format("iceberg")
+                .mode(SaveMode.Append)
+                .option(SparkWriteOptions.CHECK_NULLABILITY, false)
+                .option(SparkWriteOptions.CHECK_ORDERING, false)
+                .saveAsTable(tableName);
+    }
+
+    @SneakyThrows
+    private void deleteRecord(Record rec) {
+        String deleteQ = "DELETE FROM " + tableName + " WHERE ";
+        // todo: check if key is not structured
```

Comment:
Added an issue so we don't forget about it: https://github.com/conduitio-labs/conduit-connector-s3-iceberg/issues/12